File: 2.Data Structure/LinkedList/Singly Linked List/Java/SinglyLinkedList/src/LinkedList.java

public class LinkedList {
    private Node head; // Head of the linked list
    private Node tail; // Tail of the linked list
    // Constructor to initialize an empty linked list
    public LinkedList() {
        this.head = null;
    }

    // Method to insert a new node at the head of the linked list
    public void insertAtHead(int data) {
        Node newNode = new Node(data);
        newNode.next = head; // New node points to the current head
        head = newNode; // Update the head to the new node
    }

    // Method to display the elements in the linked list
    public void display() {
        // Create a temporary reference to iterate through the list
        Node temp = this.head;
        while (temp != null) {
            System.out.print("[data : "+temp.getData()+" | next : "+temp.next+"] --->"); // Print the data of the current node
            temp = temp.next; // Move to the next node
        }
    }
}

File: 2.Data Structure/LinkedList/Singly Linked List/Javascript/Singly/LinkedList.js
// 10 --> 5 --> 16

/* 
let myLinkedList = {
  head: {
    value: 10,
    next: {
      value: 5,
      next: {
        value: 16,
        next: null
      },
    },
  },
};
*/

class Node {
  constructor(value) {
    (this.value = value), (this.next = null);
  }
}

class LinkedList {
  constructor(value) {
    this.head = {
      value,
      next: null,
    };
    this.tail = this.head;
    this.length = 1;
  }

  append(value) {
    const newNode = new Node(value);
    this.tail.next = newNode;
    this.tail = newNode;
    this.length++;
    return this;
  }

  prepend(value) {
    const newNode = new Node(value);
    newNode.next = this.head;
    this.head = newNode;

    if (this.tail === this.head) {
      this.tail = newNode;
    }
    this.length++;
    return this;
  }

  insert(index, value) {
    if (index >= this.length) {
      this.append(value);
    } else {
      const newNode = new Node(value);
      const leader = this.traverseToIndex(index - 1);
      const holdingPointer = leader.next;
      leader.next = newNode;
      newNode.next = holdingPointer;
      this.length++;
    }
    return this;
  }

  remove(index) {
    const leader = this.traverseToIndex(index - 1);
    const unwanted = leader.next;
    leader.next = unwanted.next;
    this.length--;
    return this;
  }

  traverseToIndex(index) {
    let currentNode = this.head;
    let i = 0;
    while (i < index) {
      currentNode = currentNode.next;
      i++;
    }
    return currentNode;
  }

  printList() {
    const arr = [];
    let currentNode = this.head;
    while (currentNode !== null) {
      arr.push(currentNode.value);
      currentNode = currentNode.next;
    }
    return arr;
  }
}

const linkedList = new LinkedList();

const myLinkedList = new LinkedList(10);
myLinkedList.append(5);
myLinkedList.append(16);
myLinkedList.prepend(4);

myLinkedList.insert(60, 20);

myLinkedList.insert(2, 20);
myLinkedList.remove(2);
const array = myLinkedList.printList();
console.log(array);
